edk2toolext/perf/fpdt_parser.py

```diff
@@ -0,0 +1,1667 @@
+        guid_uint32 (int): The first 32 bits of the GUID.
+        guid_uint16_0 (int): The first 16-bit segment of the GUID.
+        guid_uint16_1 (int): The second 16-bit segment of the GUID.
+        guid_uint8_0 to guid_uint8_7 (int): The 8 individual bytes of the GUID.
+
+    Methods:
+        to_xml(): Converts the event record to an XML representation.
+    """
+
+    struct_format = "=HIQIHHBBBBBBBB"
+    size = struct.calcsize(struct_format)
+
+    def __init__(
+        self,
+        record_header: FbptRecordHeader,
+        contents_byte_array: bytes,
+        string_byte_array: bytes,
+        string_size: int,
+    ) -> None:
+        """Initializes a DynamicStringEventRecord instance.
+
+        Args:
+            record_header (FbptRecordHeader): The header of the FBPT record.
+            contents_byte_array (bytes): Byte array containing the contents of the record.
+            string_byte_array (bytes): Byte array containing the string data.
+            string_size (int): The size of the string in bytes.
+        """
+        self.header = record_header
+        string_format = f"={string_size}s"
+        self.string = struct.unpack_from(string_format, string_byte_array)[0]
+        (
+            self.progress_id,
+            self.apic_id,
+            self.timestamp,
+            self.guid_uint32,
+            self.guid_uint16_0,
+            self.guid_uint16_1,
+            self.guid_uint8_0,
+            self.guid_uint8_1,
+            self.guid_uint8_2,
+            self.guid_uint8_3,
+            self.guid_uint8_4,
+            self.guid_uint8_5,
+            self.guid_uint8_6,
+            self.guid_uint8_7,
+        ) = struct.unpack_from(DynamicStringEventRecord.struct_format, contents_byte_array)
+        try:
+            self.string = self.string[: self.string.index(b"\x00")]
+        except ValueError:
+            logging.critical("String in a Dynamic String Record does not contain a null terminator")
+        self.string = "".join(
+            [i if ((ord(i) == 0) or ((ord(i) > 31) and (ord(i) < 127))) else "?" for i in self.string.decode("ascii")]
+        )
+
+    def __str__(self) -> str:
+        """Generate a string representation of the object.
+
+        Returns:
+            str: A formatted string representation of the object's data.
+        """
+        return (
+            f"{self.header}\n"
+            f"""  Dynamic String Event Record Contents
+------------------------------------------------------------------
+  Progress ID : 0x{self.progress_id:04X}
+  Apic ID     : 0x{self.apic_id:08X}
+  Timestamp   : 0x{self.timestamp:016X}
+  GUID        : {self.guid_uint32:08X}-{self.guid_uint16_0:04X}-{self.guid_uint16_1:04X}-"""
+            f"""{self.guid_uint8_0:02X}{self.guid_uint8_1:02X}-{self.guid_uint8_2:02X}{self.guid_uint8_3:02X}"""
+            f"""{self.guid_uint8_4:02X}{self.guid_uint8_5:02X}{self.guid_uint8_6:02X}{self.guid_uint8_7:02X}
+  String      : {self.string.split("\x00")[0]}
```

Comment:
In a fstring starting with a `"`, you cannot use double quotes in the `{...}` replacement. I think switching it to single quotes (\`\x00\`) should work. This was not added until recently, so it is failing the 3.10 MSV tests.

Reply:
Download 3.10 and updated these.